### PR TITLE
JW-465, JW-466: pkt0 pipe handling

### PR DESCRIPTION
**Pull request by sodar**

### CODE REVIEW

File: windows/vrouter_mod.c [hunk 1]
```diff
@@ -200,16 +200,32 @@ SxExtUninitializeVRouter(struct vr_switch_context* ctx)
...
+
+    if (ctx->ksync_up)
         KsyncDestroyDevice(SxDriverObject);
 }
 
 NDIS_STATUS
 SxExtInitializeVRouter(struct vr_switch_context* ctx)
 {
-    if (ctx->vrouter_up || ctx->device_up || ctx->message_up || ctx->assoc_up)
+    if (ctx->vrouter_up || ctx->ksync_up || ctx->pkt0_up || ctx->device_up || ctx->message_up || ctx->assoc_up)
```
Comment (michal-clapinski) [inline windows/vrouter_mod.c]: This whole line is weird. Why is it even here? I'd put an assert here instead.

File: include/vr_windows.h [hunk 2]
```diff
@@ -31,6 +31,8 @@ struct vr_switch_context {
     BOOLEAN                 restart;
 
     BOOLEAN                 vrouter_up;
+    BOOLEAN                 ksync_up;
+    BOOLEAN                 pkt0_up;
```
Comment (michal-clapinski) [inline include/vr_windows.h]: I know these is not a part of this PR but if you're going to commit again could you put all of those in order they have to be initialized, please?

File: windows/vr_devices.c [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+VRouterUninitializeDevices(PDRIVER_OBJECT DriverObject)
+{
+    UNREFERENCED_PARAMETER(DriverObject);
+}
+
+NTSTATUS
+VRouterSetUpNamedPipeServer(_In_ PDRIVER_OBJECT DriverObject,
+                            _In_ PCWSTR DeviceName,
+                            _In_ PCWSTR DeviceSymlink,
+                            _In_ struct vr_device_dispatch_callbacks *Callbacks,
```
Comment (michal-clapinski) [inline windows/vr_devices.c]: Preferably two typedefs: ABC and PABC for pointer

Comment (sodar) [inline windows/vr_devices.c]: I agree, that is a good idea.

File: windows/vr_devices.c [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        return Status;
+    }
+
+    Status = IoCreateDevice(DriverObject, sizeof(struct vr_device_dispatch_callbacks), &_DeviceName,
+                            FILE_DEVICE_NAMED_PIPE, FILE_DEVICE_SECURE_OPEN,
+                            FALSE, &_DeviceObject);
+    if (NT_SUCCESS(Status)) {
+        *DeviceObject = _DeviceObject;
+
+        _Callbacks = (struct vr_device_dispatch_callbacks *)(*DeviceObject)->DeviceExtension;
```
Comment (sodar) [inline windows/vr_devices.c]: Yes, it can be used here. Good catch.

File: include/vr_windows.h (hunk 2 above)
Comment (sodar) [inline include/vr_windows.h]: Ok

File: windows/vrouter_mod.c (hunk 1 above)
Comment (sodar) [inline windows/vrouter_mod.c]: Yes, `assert` seems like a good idea here.

**Comment by sodar** [comment]
Retest this please

File: windows/vr_pkt0.c [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+static BOOLEAN Pkt0SymlinkCreated = FALSE;
+
+static NTSTATUS
+Pkt0DispatchCreate(PDEVICE_OBJECT DeviceObject, PIRP Irp)
+{
+    UNREFERENCED_PARAMETER(DeviceObject);
+
+    /* TODO(sodar): Implement */
+
+    IoCompleteRequest(Irp, IO_NO_INCREMENT);
```
Comment (Attil) [inline windows/vr_pkt0.c]: Are you sure calling `IoCompleteRequest` without setting `IoStatus` is valid?

File: include/vr_devices.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+typedef struct _VR_DEVICE_DISPATCH_CALLBACKS *PVR_DEVICE_DISPATCH_CALLBACKS;
+
+NTSTATUS KsyncCreateDevice(PDRIVER_OBJECT DriverObject);
+VOID KsyncDestroyDevice(PDRIVER_OBJECT DriverObject);
+
+NTSTATUS Pkt0CreateDevice(PDRIVER_OBJECT DriverObject);
+VOID Pkt0DestroyDevice(PDRIVER_OBJECT DriverObject);
+
+NTSTATUS VRouterInitializeDevices(PDRIVER_OBJECT DriverObject);
+VOID VRouterUninitializeDevices(PDRIVER_OBJECT DriverObject);
```
Comment (Attil) [inline include/vr_devices.h]: Is this function necessary? If both devices are handled on their own, there is no need for congregate function, is there?

**Comment by Attil** [comment]
Feel free to dismiss this review, because Im not completely sure on both points.

File: include/vr_devices.h (hunk 6 above)
Comment (sodar) [inline include/vr_devices.h]: `VRouterInitializeDevices` is used to update `MajorFuncion` table which is used by both pipes. `VRouterUninitializeDevices` does not do anything meaningful, it is there only to preserve symmetry in init/deinit code in `vrouter_mod.c`.

File: windows/vr_pkt0.c (hunk 5 above)
Comment (sodar) [inline windows/vr_pkt0.c]: It seems that it is not. I will add it.

**Comment by sodar** [comment]
Retest this please

**Comment by sodar** [comment]
Retest this please

**Pull request merged**